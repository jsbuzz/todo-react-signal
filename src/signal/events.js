import { Event } from '../../react-signal/event-hive/event';

export const AddTodoEvent = class extends Event {
  constructor(title) {
    super();
    this.title = title;
  }
};
export const IdEvent = class extends Event {
  constructor(id) {
    super();
    this.id = id;
  }
};

export const ErrorEvent = class extends Event {
  constructor(error) {
    super();
    this.error = error;
  }
};

export const AddTodo = AddTodoEvent.withAlias('Todo:add');
export const RemoveTodo = IdEvent.withAlias('Todo:remove');
export const ToggleTodo = IdEvent.withAlias('Todo:toggle');
