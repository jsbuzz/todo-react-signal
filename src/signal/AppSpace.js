import {
  NameSpace,
  InitState,
  set,
} from '../../react-signal/event-hive/namespace';
import { UpdateActive } from './events';

const AppSpace = NameSpace.schema('Ns.App', () => ({
  activeTodos: [
    InitState, set(0),
    
    // update number of active todos
    UpdateActive, activeTodos => ({ active }) => console.log(active) || active
  ]
}));

window.AppSpace = AppSpace;

export default AppSpace;
