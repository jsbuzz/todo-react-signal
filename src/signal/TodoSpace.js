import {
  NameSpace,
  InitState,
  set,
  modify
} from '../../react-signal/event-hive/namespace';
import { AddTodo, RemoveTodo, UpdateTodo, RestoreTodos } from './events';

const TodoSpace = NameSpace.schema((todoId = 0) => ({
  todos: [
    InitState, set(new Map()),
    
    // Add new todo
    AddTodo, modify(todos => ({ title }) => {
      const id = ++todoId;
      const todo = { id, title, done: false };

      todos.set(id, todo);
    }),
    
    // restore todos from cache
    RestoreTodos, modify(todos => ({ savedTodos }) => {
      savedTodos.forEach(todo => {
        todoId = todo.id;
        todos.set(todo.id, todo);
      });
    }),
    
    // remove a todo by id
    RemoveTodo, modify(todos => ({ id }) => todos.delete(id)),
    
    // update a todo by id
    UpdateTodo, modify(todos => ({ todo }) => todos.set(todo.id, todo))
  ]
}));

window.TodoSpace = TodoSpace;

export default TodoSpace;
