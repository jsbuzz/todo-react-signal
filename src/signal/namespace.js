import { NameSpace, InitState } from '../../react-signal/event-hive/namespace';
import { AddTodo, RemoveTodo, ToggleTodo } from './events';

const TodoSpace = NameSpace.get('Ns.Todos');

const findTodo = (todos, todoId) => {
  let i = 0;
  for (const todo of todos) {
    if (todo.id === todoId) {
      return i;
    }
    i += 1;
  }
  return -1;
};

const toggleTodo = (todo) => {
  todo.done = !todo.done;
}

let todoId = 0;
TodoSpace.defineState({
  todos: state => [
    InitState, () => (state.todos = []),
    AddTodo, ({ title }) => (state.todos.push({
      id: ++todoId,
      title,
      done: false,
    })),
    RemoveTodo, ({ id }) => (state.todos.splice(findTodo(state.todos, id), 1)),
    ToggleTodo, ({ id }) => (toggleTodo(state.todos[findTodo(state.todos, id)])),
  ],
});

export default TodoSpace;
