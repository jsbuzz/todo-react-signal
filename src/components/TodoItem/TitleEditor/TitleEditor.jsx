import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';

import { UpdateTodo } from '../../../signal/events';

class TitleEditor extends PureComponent {
  inputRef = React.createRef()

  constructor(props) {
    super(props);

    const { todo } = this.props;
    this.state = { value: todo.title };
  }

  componentDidMount() {
    this.setFocus();
  }

  onChange = (event) => {
    this.setState({ value: event.target.value });
  }

  onKeyUp = ({ keyCode }) => {
    const { todo } = this.props;
    if (keyCode === 13) {
      const { value } = this.state;
      this.namespace().trigger(UpdateTodo.with({
        ...todo,
        edited: false,
        title: value,
      }));
    } else if (keyCode === 27) {
      this.reset();
    }
  }

  setFocus() {
    this.inputRef.current.focus();
  }

  reset = () => {
    const { todo } = this.props;
    this.namespace().trigger(UpdateTodo.with({
      ...todo,
      edited: false,
    }));
  }

  render() {
    const { value } = this.state;
    return (
      <input
        type="text"
        value={value}
        onChange={this.onChange}
        onKeyUp={this.onKeyUp}
        onBlur={this.reset}
        className="edit-title"
        ref={this.inputRef}
      />
    );
  }
}

TitleEditor.propTypes = {
  todo: PropTypes.shape({
    id: PropTypes.number.isRequired,
    title: PropTypes.string.isRequired,
    done: PropTypes.bool.isRequired,
  }).isRequired,
};


export default TitleEditor;
