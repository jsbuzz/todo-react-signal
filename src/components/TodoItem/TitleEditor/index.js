import Connect from '../../../../react-signal';
import TitleEditor from './TitleEditor';

export default Connect(TitleEditor);
