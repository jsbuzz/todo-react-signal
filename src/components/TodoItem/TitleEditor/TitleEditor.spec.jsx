import React from 'react';
import { shallow } from 'enzyme';
import Connect, { TestNameSpace } from '../../../../react-signal/test-connect';

import { UpdateTodo } from '../../../signal/events';
import TitleEditor from './TitleEditor';

describe('components/TitleEditor', () => {
  const todo = {
    id: 191,
    title: 'test me!',
    done: false,
    edited: true
  };
  let NS;
  let component;

  beforeEach(() => {
    NS = new TestNameSpace();
    const ConnectedComponent = Connect(TitleEditor, NS);
    component = shallow(<ConnectedComponent todo={todo} />);
  });

  it('should match snapshot', () => {
    expect(component).toMatchSnapshot();
  });

  describe('WHEN `Enter` key is pressed', () => {
    beforeEach(() => {
      component.setState({ value: '' });
      component.find('input').simulate('keyUp', { keyCode: 13 });
    });

    describe('AND current value is empty', () => {
      it('should NOT trigger UpdateTodo', () => {
        expect(NS.lastEvent).toBe(null);
      });
    });

    describe('AND current value is NOT empty', () => {
      const newTitle = 'The new title';

      beforeEach(() => {
        component.setState({ value: newTitle });
        component.find('input').simulate('keyUp', { keyCode: 13 });
      });

      it('should trigger UpdateTodo', () => {
        expect(NS.lastEvent).toBeInstanceOf(UpdateTodo);
        expect(NS.lastEvent.todo).toEqual({
          ...todo,
          edited: false,
          title: newTitle
        });
      });
    });
  });

  describe('WHEN `Esc` key is pressed', () => {
    const newTitle = 'The new title';

    beforeEach(() => {
      component.setState({ value: newTitle });
      component.find('input').simulate('keyUp', { keyCode: 27 });
    });

    it('should trigger UpdateTodo with edited: false and original title', () => {
      expect(NS.lastEvent).toBeInstanceOf(UpdateTodo);
      expect(NS.lastEvent.todo).toEqual({
        ...todo,
        edited: false
      });
    });
  });

  describe('WHEN input loses focus', () => {
    const newTitle = 'The new title';

    beforeEach(() => {
      component.setState({ value: newTitle });
      component.find('input').simulate('blur');
    });

    it('should trigger UpdateTodo with edited: false and original title', () => {
      expect(NS.lastEvent).toBeInstanceOf(UpdateTodo);
      expect(NS.lastEvent.todo).toEqual({
        ...todo,
        edited: false
      });
    });
  });
});
