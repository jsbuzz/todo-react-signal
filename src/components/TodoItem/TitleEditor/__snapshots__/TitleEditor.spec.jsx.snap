// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/TitleEditor should match snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TitleEditor
    todo={
      Object {
        "done": false,
        "edited": true,
        "id": 191,
        "title": "test me!",
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "className": "edit-title",
      "onBlur": [Function],
      "onChange": [Function],
      "onKeyUp": [Function],
      "type": "text",
      "value": "test me!",
    },
    "ref": Object {
      "current": null,
    },
    "rendered": null,
    "type": "input",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "className": "edit-title",
        "onBlur": [Function],
        "onChange": [Function],
        "onKeyUp": [Function],
        "type": "text",
        "value": "test me!",
      },
      "ref": Object {
        "current": null,
      },
      "rendered": null,
      "type": "input",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
